### ISSUE TITLE
What attention format should I use with flash attention's qkvpacked?

### ISSUE THREAD

**Issue by vedantroy** [NVIDIA/TransformerEngine#1110]
I'm currently using `flash_attn_varlen_qkvpacked_func`, which accepts a `qkv` tensor with the format `(total, 3, nheads, headdim)`, where total = total number of tokens in the batch.

I want to switch to using the `DotProductAttention` class from this library. But, I'm not sure which `qkv_format` to use. From reading the docs, it looks like I should use `thd`. But, I have a few questions:
1. The docs say `t` is the total number of sequences in the batch. I assume this is a mistake and it should actually be the total number of tokens in the batch?
2. The asserts that if using `thd` format, the attention mask type must be `padding` or `padding_casual`. This confuses me. I don't want any attention mask, since I'm doing non-causal attention. How should I setup my kwargs for this?
3. Do I need to pass in `cu_seqlens_q_padded`, if I'm using the `thd` format?

Is there an example of using `DotProductAttention`, so I can look at that for inspo?

**Comment by cyanguwa** [comment]
Hi @vedantroy ,

We have some documentation [here](https://github.com/NVIDIA/TransformerEngine/blob/main/docs/examples/attention/attention.ipynb), and some unit tests [here](https://github.com/NVIDIA/TransformerEngine/blob/941364df3e6f6c00b4aace90024e9492eb09b511/tests/pytorch/fused_attn/test_fused_attn.py#L230) and [here](https://github.com/NVIDIA/TransformerEngine/blob/941364df3e6f6c00b4aace90024e9492eb09b511/tests/pytorch/fused_attn/test_fused_attn.py#L642) that you can use as templates.

1. I've just fixed the typos in our docstring related to the `t` in `thd (#1111). Thanks for catching that!
2. If you're doing non-causal attention, you can use `qkv_format=thd` and `attn_mask_type=padding`. You do need to specify `cu_seqlens_q` and `cu_seqlens_kv` though if you're not providing an `attention_mask` tensor. This is to help the backend to figure out how long exactly each sequence is.
3. If you don't have any padding between sequences in a batch, i.e. `[a, a, a, b, b, c, c, c, c]` rather than `[a, a, a, PAD, b, b, PAD, c, c, c, c]` for a batch of 3 sequences, you do not need to specify `cu_seqlens_q_padded` or `cu_seqlens_kv_padded`. You can take a look at [this](https://github.com/NVIDIA/TransformerEngine/blob/941364df3e6f6c00b4aace90024e9492eb09b511/tests/pytorch/fused_attn/test_fused_attn.py#L650) test in particular.

Hope this helps!

Thanks,
Charlene

**Comment by vedantroy** [comment]
Thanks for the response! I'm assuming the sequence lengths are cumulative?

**Comment by cyanguwa** [comment]
Yes, `cu_seqlens_q` and `cu_seqlens_kv` are.
